src/tests/integration/IntegrationBase.t.sol

```diff
@@ -0,0 +1,938 @@
+        assertEq(ink, 0);
+        assertEq(art, 0);
+        assertEq(vat.gem(ilk, address(pool)), 0);
+        assertEq(vat.gem(ilk, address(end)), pink);
+        assertGe(getLPTokenBalanceInAssets(address(pool)), pink);
+        if (originalSin + part * RAY >= originalDai) {
+            assertApproxEqAbs(vat.sin(address(vow)), originalSin + part * RAY - originalDai, RAY * roundingTolerance);
+            assertEq(vat.dai(address(vow)), 0);
+        } else {
+            assertApproxEqAbs(vat.dai(address(vow)), originalDai - originalSin - part * RAY, RAY * roundingTolerance);
+            assertEq(vat.sin(address(vow)), 0);
+        }
+
+        // We try to unwind what is possible
+        hub.exec(ilk);
+        vow.heal(_min(vat.sin(address(vow)), vat.dai(address(vow))));
+
+        // A part can't be unwind yet
+        assertEq(vat.gem(ilk, address(end)), amountSupplied / 2);
+        assertGe(getLPTokenBalanceInAssets(address(pool)), amountSupplied / 2);
+        if (originalSin + part * RAY >= originalDai + (amountSupplied / 2) * RAY) {
+            // rounding may affect twice, and multiplied by RAY to be compared with sin
+            assertApproxEqAbs(vat.sin(address(vow)), originalSin + part * RAY - originalDai - (amountSupplied / 2) * RAY, 2 * RAY * roundingTolerance);
+            assertEq(vat.dai(address(vow)), 0);
+        } else {
+            // rounding may affect twice, and multiplied by RAY to be compared with sin
+            assertApproxEqAbs(vat.dai(address(vow)), originalDai + (amountSupplied / 2) * RAY - originalSin - part * RAY, 2 * RAY * roundingTolerance);
+            assertEq(vat.sin(address(vow)), 0);
+        }
+
+        // Then pool gets some liquidity
+        adjustLiquidity(int256(amountToBorrow));
+
+        // Rest of the liquidity can be withdrawn
+        hub.exec(ilk);
+        vow.heal(_min(vat.sin(address(vow)), vat.dai(address(vow))));
+        assertEq(vat.gem(ilk, address(end)), 0);
+        assertEq(getLPTokenBalanceInAssets(address(pool)), 0);
+        assertEq(vat.sin(address(vow)), 0);
+        assertApproxEqAbs(vat.dai(address(vow)), originalDai - originalSin + daiEarned * RAY, RAY * roundingTolerance);
+    }
+
+    function test_uncull_not_culled() public {
+        adjustDebt(standardDebtSize);
+        vm.prank(admin); hub.cage(ilk);
+
+        // MCD shutdowns
+        vm.prank(admin); end.cage();
+
+        assertRevert(address(hub), abi.encodeWithSignature("uncull(bytes32)", ilk), "D3MHub/not-prev-culled");
+    }
+
+    function test_uncull_not_shutdown() public {
+        adjustDebt(standardDebtSize);
+        vm.prank(admin); hub.cage(ilk);
+
+        (, , uint256 tau, , ) = hub.ilks(ilk);
+        vm.warp(block.timestamp + tau);
+
+        hub.cull(ilk);
+
+        assertRevert(address(hub), abi.encodeWithSignature("uncull(bytes32)", ilk), "D3MHub/no-uncull-normal-operation");
+    }
+
+    function test_cage_exit() public {
+        adjustDebt(100 ether);
```

Comment:
So maybe the main difference of this test vs the one in V2 is that the `totalArt` will match the amount of gems we are giving to the user. Maybe I'd just increase the buffer to something else, so we can prove the checks still hold if we are redeeming a portion (which is a more realistic case).

Reply:
fixed